nix/clickhouse/default.nix

```diff
@@ -0,0 +1,165 @@
+        of Postgres. The schema attribute is optional: If not specified, an empty database is created.
+      '';
+      example = lib.literalExpression ''
+        [
+          {
+            name = "foodatabase";
+            schemas = [ ./fooschemas ./bar.sql ];
+          }
+          { name = "bardatabase"; }
+        ]
+      '';
+    };
+    outputs.settings = lib.mkOption {
+      type = types.deferredModule;
+      internal = true;
+      readOnly = true;
+      default = {
+        processes =
+          let
+            clickhouseConfig = pkgs.writeText "clickhouse-config.yaml" ''
+              logger:
+                level: warning
+                console: 1
+              tcp_port: ${toString config.port}
+              default_profile: default
+              default_database: default
+              path: ${config.dataDir}/clickhouse
+              tmp_path: ${config.dataDir}/clickhouse/tmp
+              user_files_path: ${config.dataDir}/clickhouse/user_files
+              format_schema_path: ${config.dataDir}/clickhouse/format_schemas
+              user_directories:
+                users_xml:
+                  path: ${config.package}/etc/clickhouse-server/users.xml
+              ${config.extraConfig}
+            '';
+          in
+          {
+            # DB initialization
+            "${name}-init" =
+              let
+                # https://github.com/ClickHouse/ClickHouse/issues/4491
+                setupInitialSchema = schema: '' < ${schema} tr -s '\r\n' ' ' | clickhouse-client -mn --port ${builtins.toString config.port}; '';
+                setupInitialDatabases =
+                  lib.concatMapStrings
+                    (database: ''
+                      echo "Creating database: ${database.name}"
+                      clickhouse-client --port ${builtins.toString config.port} --query "CREATE DATABASE iF NOT EXISTS ${database.name}"
+                      echo "Database successfully created: ${database.name}"
+                      ${lib.optionalString (database.schemas != null)
+                        (lib.concatMapStrings (schema: setupInitialSchema schema) database.schemas)}
+                    '')
+                    config.initialDatabases;
+                setupScript = pkgs.writeShellApplication {
+                  name = "setup-clickhouse";
+                  runtimeInputs = with pkgs; [ config.package coreutils gnugrep gawk ];
+                  text = ''
+                    if test -d ${config.dataDir}
+                      then echo "Clickhouse database directory ${config.dataDir} appears to contain a database; Skipping initialization"
+                      else
+                        echo "Clickhouse is setting up the initial database."
+                        set -m
+                        clickhouse-server --config-file=${clickhouseConfig} &
+                        sleep 5s
```

Comment:
Starting `clickhouse-server` in background with `&` and waiting for 5s, which we aren’t even sure if its going to be enough, isn’t a good way to initialise. Although I tried a couple of things:
- `clickhouse-server —daemon`: doesn’t seem to work at all and I have no clue why.
- `clickhouse start` should start in background according to [docs](https://clickhouse.com/docs/en/install#launch) but it expects sudo as it tries to create `/var/run/clickhouse-server` and I don’t see any way to override this.

As I don’t see any better approach for now, we can keep this implementation as is and visit this in future.

Reply:
> Starting clickhouse-server in background with & and waiting for 5s, which we aren’t even sure if its going to be enough, isn’t a good way to initialise.
anyway if 5 secs will be not enough, then init script will fail and main process will not start, and we will face error. I've tested several times, and all worked fine

Reply:
Yes, that’s true but some machines might be too slow to even initialise in 5 seconds or someone might use a configuration that might cause delay in initialisation. Anyway, we can look at it if that ever happens.